### ISSUE TITLE
o.a.n.r.NettyRequestSender$1.operationComplete() java.lang.NullPointerException: timeoutsHolder

### ISSUE THREAD

**Issue by Spikhalskiy** [AsyncHttpClient/async-http-client#1180]
Hi,

I'm getting the following exception with AsyncHttpClient sporadically

```
[06 Jun 2016 13:02:35,198][WARN ][DefaultPromise,collectedRequestDisruptor73689328] - An exception was thrown by org.asynchttpclient.netty.request.NettyRequestSender$1.operationComplete()
java.lang.NullPointerException: timeoutsHolder
	at org.asynchttpclient.util.Assertions.assertNotNull(Assertions.java:23)
	at org.asynchttpclient.netty.request.NettyChannelConnector.<init>(NettyChannelConnector.java:54)
	at org.asynchttpclient.netty.request.NettyRequestSender$1.onSuccess(NettyRequestSender.java:281)
	at org.asynchttpclient.netty.request.NettyRequestSender$1.onSuccess(NettyRequestSender.java:276)
	at org.asynchttpclient.netty.SimpleFutureListener.operationComplete(SimpleFutureListener.java:24)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:683)
	at io.netty.util.concurrent.DefaultPromise.notifyLateListener(DefaultPromise.java:624)
	at io.netty.util.concurrent.DefaultPromise.addListener(DefaultPromise.java:139)
	at io.netty.util.concurrent.DefaultPromise.addListener(DefaultPromise.java:33)
	at org.asynchttpclient.netty.request.NettyRequestSender.sendRequestWithNewChannel(NettyRequestSender.java:276)
	at org.asynchttpclient.netty.request.NettyRequestSender.sendRequestWithCertainForceConnect(NettyRequestSender.java:136)
	at org.asynchttpclient.netty.request.NettyRequestSender.sendRequest(NettyRequestSender.java:107)
	at org.asynchttpclient.DefaultAsyncHttpClient.execute(DefaultAsyncHttpClient.java:220)
	at org.asynchttpclient.DefaultAsyncHttpClient.executeRequest(DefaultAsyncHttpClient.java:188)
	at org.asynchttpclient.BoundRequestBuilder.execute(BoundRequestBuilder.java:35)
        ...
```

Version: 2.0.4

Any ideas?

**Comment by Spikhalskiy** [comment]
May be related
```
[06 Jun 2016 13:30:21,059][ERROR][ResourceLeakDetector,AsyncHttpClient-2-13] - LEAK: ByteBuf.release() was not called before it's garbage-collected. Enable advanced leak reporting to find out where the leak occurred. To enable advanced leak reporting, specify the JVM option '-Dio.netty.leakDetection.level=advanced' or call ResourceLeakDetector.setLevel() See http://netty.io/wiki/reference-counted-objects.html for more information.
```

**Comment by Spikhalskiy** [comment]
About Leak - I'm not sure it's useful to put here full output, but I analyzed it and what I found. I have class

```
public class ReleasingByteBufInputStream extends ByteBufInputStream {
    private ByteBuf buffer;

    public ReleasingByteBufInputStream(ByteBuf buffer) {
        super(buffer);
        ReferenceCountUtil.retain(buffer);
        this.buffer = buffer;
    }

    @Override
    public void close() throws IOException {
        super.close();
        ReferenceCountUtil.release(buffer);
        buffer = null; //to make close idempotent
    }
}
```

Which I use this way:

```
    public void onEvent(CollectedEvent event, long sequence, boolean endOfBatch) throws Exception {
        try {
            if (event.isSend()) {
                send(event);
            }
        } finally {
            ReferenceCountUtil.release(event.getFullHttpRequest());
        }
    }

    private void send(CollectedEvent event) {
        httpClient.preparePost(endpointUrl)
                .setHeaders(adoptHeaders(fullHttpRequest.headers()))
                .setBody(new ReleasingByteBufInputStream(fullHttpRequest.content()))
                .execute(handler);
    }
```
And there is no call to ReferenceCountUtil.release from ReleasingByteBufInputStream#close() in netty's log. So, looks like, in some cases AsyncHttpClient doesn't call close on passed input stream. Could be related to subj problem.

**Comment by slandelle** [comment]
Indeed, a body `InputStream` is only closed once it has been consumed, even partially. It's typically not closed when the connection could not be established so that the user can retry.
Have you tried releasing in your handler when the request is not successful?

**Comment by Spikhalskiy** [comment]
@slandelle No, I didn't. It could be a reason. Ugh, pretty unobvious. Could this stuff be made configurable? Don't want to pass too much additionally to handler entity.

**Comment by Spikhalskiy** [comment]
Reproducer could be hard to prepare, but if you have no idea just with an exception - I would try.

**Comment by slandelle** [comment]
> Could this stuff be made configurable?

Not atm. Maybe once I have enough feedback regarding streams usage. Some day...

**Comment by slandelle** [comment]
I suspect this NPE happens when the Future gets cancelled (typically a timeout) while hostname is being resolved.

**Comment by Golizart** [comment]
Hi,
I'm getting the following exception with AsyncHttpClient sporadically
`2021-01-13 13:40:37,374 WARN i.n.u.c.DefaultPromise [AsyncHttpClient-4-29] An exception was thrown by org.asynchttpclient.netty.channel.NettyChannelConnector$1.operationComplete()
java.lang.NullPointerException: null
	at org.asynchttpclient.netty.channel.NettyConnectListener.onSuccess(NettyConnectListener.java:105) ~[async-http-client-2.10.4.jar:?]
	at org.asynchttpclient.netty.channel.NettyChannelConnector$1.onSuccess(NettyChannelConnector.java:92) ~[async-http-client-2.10.4.jar:?]
	at org.asynchttpclient.netty.SimpleChannelFutureListener.operationComplete(SimpleChannelFutureListener.java:26) ~[async-http-client-2.10.4.jar:?]
	at org.asynchttpclient.netty.SimpleChannelFutureListener.operationComplete(SimpleChannelFutureListener.java:20) ~[async-http-client-2.10.4.jar:?]
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577) [netty-common-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:570) [netty-common-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:549) [netty-common-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490) [netty-common-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615) [netty-common-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.concurrent.DefaultPromise.setSuccess0(DefaultPromise.java:604) [netty-common-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:104) [netty-common-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.fulfillConnectPromise(AbstractNioChannel.java:300) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:335) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:702) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) [netty-transport-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.45.Final.jar:4.1.45.Final]
	at java.lang.Thread.run(Thread.java:834) [?:?]`

Version: 2.10.4
 I think this fix(40f89d5) didn't solve the problem
